[CAP-2461] Update CustomResource config with complex types (#30356)
DataDog/documentation: content/en/infrastructure/containers/configuration.md

1. On the modal, under **Indexing Configuration**, select the fields you want to index from the custom resource.

    {{< img src="infrastructure/containers_view/CRD_indexing_2.mp4" alt="A video of the Collecting and Indexing modal. The cursor selects three fields and clicks Enable Indexing. A success message displays." video="true">}}

#### Indexing complex types

For arrays of objects, two indexing strategies are available:

> -   **Field**: Groups object array nested fields under each unique value of the selected field.
> -   **No field**: Flattens object array nested fields into arrays, merging all field values.

Reviewer: Curious if this could be simplified a bit to improve readability. Quick suggestion (please feel free to modify it)
```suggestion
-   **Field**: Object's nested fields are indexed based on each unique value of Field.
-   **No field**: Object's nested fields are indexed solely on nested field name.
```

Author: Agreed, updated here: a945fa2996763705e9d8bf69fdb09dcfcce9a7ca

{{< tabs >}}
{{% tab "CR1" %}}

```yaml
status:
    conditions:
        - type: HorizontalAbleToScale
          status: 'True'
        - type: VerticalAbleToApply
          status: 'False'
```

{{% /tab %}}
{{% tab "CR2" %}}

```yaml
status:
    conditions:
        - type: VerticalAbleToApply
          status: 'True'
        - type: HorizontalAbleToScale
          status: 'False'
```

{{% /tab %}}
{{< /tabs >}}

You have the filtering possibilities on `status.conditions` based on the two indexing strategies:

{{< tabs >}}
{{% tab "Grouping by no field" %}}

**Indexed fields for CR1:**

```yaml
status:
    conditions:
        type: [HorizontalAbleToScale, VerticalAbleToApply]
        status: ['True', 'False']
```

**Indexed fields for CR2:**

```yaml
status:
    conditions:
        type: [VerticalAbleToApply, HorizontalAbleToScale]
        status: ['True', 'False']
```

**Example queries:**

**Query 1:**

```text
field#status.conditions.status:"False"
```

**Result:** Returns CR1 and CR2. Both CRs have at least one object with `status:"False"`

**Query 2:**

```text
field#status.conditions.status:"False" AND field#status.conditions.type:VerticalAbleToApply
```

> **Result:** Returns CR1 and CR2. Each filter is evaluated separately. At least one `status.condition` object in each custom resource matches each filter, even if it's not the same object that matches both filters.

Reviewer: >Each filter is evaluated separately

This makes me think that the queries `field#status.conditions.status:"False"` and `field#status.conditions.type:VerticalAbleToApply` are evaluated separately in the backend which is not true.

```suggestion
**Result:** Returns CR1 and CR2. At least one `status.condition` object in each custom resource matches one of the filters - even if it's not the same object that matches both filters.
```

{{% /tab %}}
{{% tab "Grouping by type" %}}

**Indexed fields for CR1:**

```yaml
status:
    conditions:
        - HorizontalAbleToScale:
              status: 'True'
        - VerticalAbleToApply:
              status: 'False'
```

**Indexed fields for CR2:**

```yaml
status:
    conditions:
        - VerticalAbleToApply:
              status: 'True'
        - HorizontalAbleToScale:
              status: 'False'
```

**Example query:**

```text
field#status.conditions.HorizontalAbleToScale.status:"False"
```

> **Result:** Returns only CR2. The query filters for CRs where `status.condition` objects of type `HorizontalAbleToScale` have `status:"False"`. Both the `type` and `status` filters must match the same object.

Author: Agreed, updated here: a945fa2996763705e9d8bf69fdb09dcfcce9a7ca

{{% /tab %}}
{{< /tabs >}}

<div class="alert alert-info">You can select up to 50 fields per resource. You can use the preview to validate your indexing choices.</div>

> Select **Enable Indexing** to save.

Reviewer: Should this be a separate step, i.e. Step 5, following this one?

<img width="762" height="71" alt="image" src="https://github.com/user-attachments/assets/819a2b0b-0d67-4b81-8207-c144f1c453d1" />
